garage/tf/regressors/base2.py

```diff
@@ -0,0 +1,184 @@
+
+        Args:
+            xs: Input data.
+
+        Return:
+            The predicted ys.
+        """
+        raise NotImplementedError
+
+    def get_params_internal(self, **tags):
+        """
+        Get the list of parameters.
+
+        This internal method does not perform caching, and should
+        be implemented by subclasses.
+        """
+        raise NotImplementedError
+
+    def get_params(self, **tags):
+        """
+        Get the list of parameters, filtered by the provided tags.
+
+        Some common tags include 'regularizable' and 'trainable'
+        """
+        tag_tuple = tuple(sorted(list(tags.items()), key=lambda x: x[0]))
+        if tag_tuple not in self._cached_params:
+            self._cached_params[tag_tuple] = self.get_params_internal(**tags)
+        return self._cached_params[tag_tuple]
+
+    def get_param_dtypes(self, **tags):
+        """Get the list of dtypes for the parameters."""
+        tag_tuple = tuple(sorted(list(tags.items()), key=lambda x: x[0]))
+        if tag_tuple not in self._cached_param_dtypes:
+            params = self.get_params(**tags)
+            param_values = tf.get_default_session().run(params)
+            self._cached_param_dtypes[tag_tuple] = [
+                val.dtype for val in param_values
+            ]
+        return self._cached_param_dtypes[tag_tuple]
+
+    def get_param_shapes(self, **tags):
+        """Get the list of shapes for the parameters."""
+        tag_tuple = tuple(sorted(list(tags.items()), key=lambda x: x[0]))
+        if tag_tuple not in self._cached_param_shapes:
+            params = self.get_params(**tags)
+            param_values = tf.get_default_session().run(params)
+            self._cached_param_shapes[tag_tuple] = [
+                val.shape for val in param_values
+            ]
+        return self._cached_param_shapes[tag_tuple]
+
+    def get_param_values(self, **tags):
+        """Get the list of values for the parameters."""
+        params = self.get_params(**tags)
+        param_values = tf.get_default_session().run(params)
+        return flatten_tensors(param_values)
+
+    def set_param_values(self, flattened_params, name=None, **tags):
+        """Set the values for the parameters."""
+        with tf.name_scope(name, 'set_param_values', [flattened_params]):
+            debug = tags.pop('debug', False)
+            param_values = unflatten_tensors(flattened_params,
+                                             self.get_param_shapes(**tags))
+            ops = []
+            feed_dict = dict()
+            for param, dtype, value in zip(
+                    self.get_params(**tags), self.get_param_dtypes(**tags),
+                    param_values):
+                if param not in self._cached_assign_ops:
+                    assign_placeholder = tf.placeholder(
+                        dtype=param.dtype.base_dtype)
+                    assign_op = tf.assign(param, assign_placeholder)
```

Comment:
Can we also replace these with variable.load?

Reply:
We can wrap a variable.load inside a session.make_callable, will that be better than tf.assign?

Reply:
Yes I think so.